homeassistant/components/google_generative_ai_conversation/__init__.py

```diff
@@ -3,44 +3,107 @@
 
 from functools import partial
 import logging
+from pathlib import Path
 from typing import Literal
 
 from google.api_core.exceptions import ClientError
-import google.generativeai as palm
-from google.generativeai.types.discuss_types import ChatResponse
+import google.generativeai as genai
+import google.generativeai.types as genai_types
+import voluptuous as vol
 
 from homeassistant.components import conversation
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import CONF_API_KEY, MATCH_ALL
-from homeassistant.core import HomeAssistant
-from homeassistant.exceptions import ConfigEntryNotReady, TemplateError
-from homeassistant.helpers import intent, template
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import (
+    ConfigEntryNotReady,
+    HomeAssistantError,
+    TemplateError,
+)
+from homeassistant.helpers import config_validation as cv, intent, template
+from homeassistant.helpers.typing import ConfigType
 from homeassistant.util import ulid
 
 from .const import (
     CONF_CHAT_MODEL,
+    CONF_MAX_TOKENS,
     CONF_PROMPT,
     CONF_TEMPERATURE,
     CONF_TOP_K,
     CONF_TOP_P,
     DEFAULT_CHAT_MODEL,
+    DEFAULT_MAX_TOKENS,
     DEFAULT_PROMPT,
     DEFAULT_TEMPERATURE,
     DEFAULT_TOP_K,
     DEFAULT_TOP_P,
+    DOMAIN,
 )
 
 _LOGGER = logging.getLogger(__name__)
+SERVICE_GENERATE_CONTENT = "generate_content"
+CONF_IMAGE_FILENAME = "image_filename"
+
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up OpenAI Conversation."""
+
+    async def generate_content(call: ServiceCall) -> ServiceResponse:
+        """Generate content from text and optionally images."""
+        prompt_parts = [call.data[CONF_PROMPT]]
+        image_filenames = call.data[CONF_IMAGE_FILENAME]
```

Comment:
Should we call it image filename or would Gemini allow in the future any type of file to be uploaded? (Like OpenAI). If so, maybe we can call it `file_input` or something.

Reply:
Currently only images are allowed. For any other mime type the API fails with: `Failed to call service google_generative_ai_conversation.generate_content. Error generating content: 400 * GenerateContentRequest.contents[0].parts[1].inline_data.mime_type: MIME type must be image/png, image/jpeg, image/webp, image/heic, or image/heif.`
No idea if they plan on supporting other types. Let me know if we should go with `file_input` to be future proof and document that only images are currently supported.

Reply:
Let's stick with image for now. If we allow other files, we can add a second arg. It's easier to create a selector when we allow only images.